### PR TITLE
Topic/missing types in config

### PR DESCRIPTION
**Pull request by aufgang001**
It is not possible to use Integers (int) for custom types in a custom configuration class as described [here](http://actor-framework.readthedocs.io/en/latest/ConfiguringActorApplications.html?highlight=config#command-line-options-and-ini-configuration-files).

With this simple fix it works.

### CODE REVIEW

File: libcaf_core/caf/config_option.hpp [hunk 1]
```diff
@@ -205,22 +222,27 @@ class config_option_impl<T, true> : public config_option {
...
   config_reader_sink to_sink() override {
-    return [=](size_t ln, config_value& x) {
+    return [=](size_t ln, config_value& x, optional<std::ostream&> errors) {
       // the INI parser accepts all integers as int64_t
+      // and all floating point numbers as doubles
       using cfg_type =
         typename std::conditional<
-          std::is_integral<value_type>::value
-          && !std::is_same<bool, value_type>::value,
+          std::is_integral<value_type>::value &&
```
Comment (Neverlord) [inline libcaf_core/caf/config_option.hpp]: https://github.com/actor-framework/actor-framework/blob/master/CONTRIBUTING.md#breaking-statements

File: libcaf_core/src/config_option.cpp [hunk 2]
```diff
@@ -17,12 +17,29 @@
...
 
 namespace caf {
 
+const char* type_name_visitor_tbl[] {
+  "a bool", 
+  "a float", 
+  "a double",
+  "a string",
+  "an atom_value",
+  "a 8-bit integer",
```
Comment (Neverlord) [inline libcaf_core/src/config_option.cpp]: "an 8-bit..."

File: libcaf_core/test/config_option.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+#include "caf/config_option.hpp"
+#include "caf/actor_system_config.hpp"
+
+using namespace std;
+using namespace caf;
+
+
+namespace {
+
+struct fixture {
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: delete unused structs

File: libcaf_core/test/config_option.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+#define CAF_SUITE config_option
+#include "caf/test/unit_test.hpp"
+
+#include "caf/config_option.hpp"
+#include "caf/actor_system_config.hpp"
+
+using namespace std;
+using namespace caf;
+
+
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: extra linebreak

File: libcaf_core/test/config_option.cpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+template<class T>
+constexpr T init_value() { return 0; }
+
+template<class T, class U, class V> 
+constexpr T failure_value() {
+  return static_cast<V>(std::numeric_limits<U>::max()) + 2;
+}
+
+template <class T, class U> 
+std::tuple<T, std::string> run_config_option(const T& init_value,
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: parameter `init_value` shadows `init_value<T>()` (fixed if `init_value<T>()` gets renamed to `zero<T>()`)

File: libcaf_core/test/config_option.cpp [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+template <class T, class U> //only works with an integral types and double 
+void check_num_general_usage() {
+  //check positive numbers
+  std::string error_str;
+  U test_value = 5;
+  T result;
+  std::tie(result, error_str) = run_config_option(init_value<T>(),
+                                                        test_value);
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: formatting off

File: libcaf_core/test/config_option.cpp [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+  T result;
+  std::tie(result, error_str) = run_config_option(init_value<T>(),
+                                                        test_value); 
+  CAF_CHECK_EQUAL(result, test_value);
+  CAF_CHECK_EQUAL(error_str.empty(), true);
+
+  //check negative numbers
+  test_value = -5;
+  std::tie(result, error_str) = run_config_option(init_value<T>(),
+                                                     test_value);
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: formatting off

File: libcaf_core/test/config_option.cpp [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+struct fixture {
+};
+
+constexpr const char* category = "category";
+constexpr const char* name = "name";
+constexpr const char* explanation = "explanation";
+constexpr size_t line = 0;
+
+template<class T>
+constexpr T init_value() { return 0; }
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: Bad name. Why not `zero<T>()`?

File: libcaf_core/test/config_option.cpp [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+  CAF_CHECK_EQUAL(result, test_value);
+  CAF_CHECK_EQUAL(error_str.empty(), true);
+
+  //check negative numbers
+  test_value = -5;
+  std::tie(result, error_str) = run_config_option(init_value<T>(),
+                                                     test_value); 
+  if (std::numeric_limits<T>::is_signed) {
+    CAF_CHECK_EQUAL(result, test_value);
+    CAF_CHECK_EQUAL(error_str.empty(), true);
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: Quite verbose to check a boolean expression with `CAF_CHECK_EQUAL`. Why not simply `CAF_CHECK(error_str.empty());`?

File: libcaf_core/test/config_option.cpp [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+constexpr const char* name = "name";
+constexpr const char* explanation = "explanation";
+constexpr size_t line = 0;
+
+template<class T>
+constexpr T init_value() { return 0; }
+
+template<class T, class U, class V> 
+constexpr T failure_value() {
+  return static_cast<V>(std::numeric_limits<U>::max()) + 2;
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: Why +2 (as opposed to +1)? `failure_value` is not a good description of the purpose. `overflow`?

File: libcaf_core/test/config_option.cpp [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+  std::tie(result, error_str) = run_config_option(init_value<T>(),
+                                                     test_value); 
+  if (std::numeric_limits<T>::is_signed) {
+    CAF_CHECK_EQUAL(result, test_value);
+    CAF_CHECK_EQUAL(error_str.empty(), true);
+  } else {
+    CAF_CHECK_EQUAL(result, init_value<T>());
+    CAF_CHECK_EQUAL(error_str.empty(), false);
+  }
+
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: Vertical whitespace in function: https://github.com/actor-framework/actor-framework/blob/master/CONTRIBUTING.md#general

File: libcaf_core/test/config_option.cpp [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+  check_num_boundary_usage<T, U>();
+}
+
+template <class T> //intended for atoms, strings, and bools 
+void check_non_num_general_usage(const T& init_value, const T& test_value) {
+  std::string error_str;
+  T result;
+  std::tie(result, error_str) = run_config_option(init_value, test_value);
+  CAF_CHECK_EQUAL(result, test_value);
+
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: https://github.com/actor-framework/actor-framework/blob/master/CONTRIBUTING.md#general

File: libcaf_core/test/config_option.cpp [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+CAF_TEST(type_float) {
+  check_num_general_usage<float, double>();
+
+  //boundary check
+  std::string error_str;
+  float result;
+  float init_value = 0;
+  double boundary_check = static_cast<double>(
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: Why is `failure_value` (aka `overflow`) not used here?

File: libcaf_core/test/config_option.cpp [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+                                      std::numeric_limits<float>::max()) * 2;
+  std::tie(result, error_str) = run_config_option(init_value, boundary_check); 
+  float float_inf = std::numeric_limits<float>::infinity();
+  //Unit test does not compare inf values correct until now
```
Comment (Neverlord) [inline libcaf_core/test/config_option.cpp]: Until now? Still doesn't as of this PR. :)

File: libcaf_core/test/config_option.cpp (hunk 10 above)
Comment (aufgang001) [inline libcaf_core/test/config_option.cpp]: +2 is necessary as after an overflow unsigned integral numbers must differ from zero() in the tests

I added a comment to this code line

**Comment by Neverlord** [comment]
👍

**Pull request merged**